kokkos/kokkos-kernels
# implement batched serial pbtrs

This PR implements [pbtrs](https://www.netlib.org/lapack/explore-html/db/dec/group__pbtrs_ga1ccbe62287933e8bf693cc0e506d167f.html) function.

Following files are added:
1. `KokkosBatched_Pbtrs_Serial_Impl.hpp`: Internal interfaces
2. `KokkosBatched_Pbtrs_Serial_Internal.hpp`: Implementation details
3. `KokkosBatched_Pbtrs.hpp`: APIs
4. `Test_Batched_SerialPbtrs.hpp`: Unit tests for that

## Detailed description
It solves the equation `A * x = b`, where `A` is a real symmetric (complex Hermitian) positive definite band matrix `A`, where `A` is stored in a [band storage](https://www.netlib.org/lapack/lug/node124.html).
Before solving, the Cholesky factorization `A = U**H*U` or `A = L*L**H`  must be computed by Pbtrf.

Here, the matrix has the following shape.
- `A`: `(batch_count, ldab, n)`  
  On entry, the upper or lower triangle of the symmetric band
  matrix A, stored in the first KD+1 rows of the array. On exit, the triangular factor `U` or `L` from the
  Cholesky factorization `A = U**H*U` or `A = L*L**H` of the band
  matrix `A`, in the same storage format as `A`
- `B`: `(batch_count, n)`  
   On entry, it contains the n element `n` right-hand side vector `b`. On exit, the solution vectors `x`.

Example of a single batch of matrix A `n = 10`.

```bash
A
4 1 0 0 0 0 0 0 0 0 
1 4 1 0 0 0 0 0 0 0 
0 1 4 1 0 0 0 0 0 0 
0 0 1 4 1 0 0 0 0 0 
0 0 0 1 4 1 0 0 0 0 
0 0 0 0 1 4 1 0 0 0 
0 0 0 0 0 1 4 1 0 0 
0 0 0 0 0 0 1 4 1 0 
0 0 0 0 0 0 0 1 4 1 
0 0 0 0 0 0 0 0 1 4

AB (upper)
0 1 1 1 1 1 1 1 1 1 
4 4 4 4 4 4 4 4 4 4

AB (lower)
4 4 4 4 4 4 4 4 4 4
1 1 1 1 1 1 1 1 1 0
```

Parallelization would be made in the following manner. This is efficient only when 
A is given in `LayoutLeft` for GPUs and `LayoutRight` for CPUs (parallelized over batch direction).

```C++
Kokkos::parallel_for('pbtrs', 
    Kokkos::RangePolicy<execution_space> policy(0, n),
    [=](const int k) {
        auto aa = Kokkos::subview(_a, k, Kokkos::ALL(), Kokkos::ALL());
        auto bb = Kokkos::subview(_b, k, Kokkos::ALL());

        KokkosBatched::SerialPbtrs<AlgoTagType>::invoke(aa, bb);
    });
```

## Tests
1.  Confirm `A * x = b` for analytical case, where
```bash
A: [[4, 1, 0],
    [1, 4, 1],
    [0, 1, 4]]
b: [1, 1, 1]
x: [3/14, 1/7, 3/14]
```

2. Make a real (complex Hermitian) symmetric positive definite banded matrix from random and keep it in a full (called `A`) or banded storage (called `AB`). Firstly, factorize `AB` in banded storage to get `U` or `L` by `pbtrf`. Then, solve `AB * x = b` with `pbtrs` to get `x`, while keeping the original `b` in `x_ref`. Finally confirm that `A * x` is equal to `b (=x_ref)` using GEMV.
